File: apps/frontend/src/api/events.ts
import { apiRequest } from "./client";

export type InteractionEventType =
  | "recommendation_impression"
  | "movie_click"
  | "movie_like"
  | "movie_dislike"
  | "movie_rating"
  | "preference_search";

export type LogInteractionEventPayload = {
  user_id: number;
  movie_id?: number | null;
  event_type: InteractionEventType;
  event_value?: number | null;
  source?: string | null;
  rank?: number | null;
  metadata?: Record<string, unknown> | null;
};

export type LogInteractionEventResponse = {
  id: number;
  user_id: number;
  movie_id: number | null;
  event_type: string;
  event_value: number | null;
  message: string;
};

export async function logInteractionEvent(
  payload: LogInteractionEventPayload
): Promise<LogInteractionEventResponse> {
  return apiRequest<LogInteractionEventResponse>("/events", {
    method: "POST",
    body: {
      user_id: payload.user_id,
      movie_id: payload.movie_id ?? null,
      event_type: payload.event_type,
      event_value: payload.event_value ?? null,
      source: payload.source ?? null,
      rank: payload.rank ?? null,
      metadata: payload.metadata ?? {},
    },
  });
}

File: apps/backend/app/schemas/events.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class InteractionEventRequest(BaseModel):
    user_id: int
    movie_id: int | None = None
    event_type: str = Field(..., examples=["movie_click", "movie_rating"])
    event_value: float | None = None
    source: str | None = None
    rank: int | None = None
    metadata: dict[str, Any] | None = None


class InteractionEventResponse(BaseModel):
    id: int
    user_id: int
    movie_id: int | None
    event_type: str
    event_value: float | None
    message: str

    feedback_updated: bool = False
    feedback_count: int | None = None
    content_refresh_job_created: bool = False
    content_refresh_job_id: int | None = None